File: Matrix Multiplication/Python/mtx_multi_vec.py
# Undergraduate Student: Arturo Burgos
# Professor: João Rodrigo Andrade
# Federal University of Uberlândia - UFU, Fluid Mechanics Laboratory - MFLab, Block 5P, Uberlândia, MG, Brazil


# Second exercise: Consider two arbitrary Matrix NxN - by a common loop


import numpy as np  # Here I'm importing the array/matrix module from python
import time  # Here I'm importing this python module in order to calculate the precise time that the function works



n = 1500 # The numbers of elements in each matrix dimension

A = np.random.randint(150, size = (n,n))  # Here I just import numpy to create an arbitrary matrix
B = np.random.randint(150, size = (n,n))

D = np.matmul(A,B) # Actual vectorizing function that does the same as the for loop below
R = np.zeros((n,n))


# i represents lines and j represents columns. k represents the line or column 11element11 indices

def multivec(mtx1, mtx2, z, mtx3):

    mtx3 = np.matmul(A,B)

    return mtx3

t_initial = time.time()  # I set here the initial time.
R = multivec(A, B, n, R)

print('\n')
print("--- %s seconds ---\n" % (time.time() - t_initial))

print("\n\nAre both results equal to each other?\n")
comparison = (R == D).all() # Here I verify if the R matrix is equal to D matrix
print(comparison)

File: Matrix Multiplication/Julia/mtx_multi_vec.jl
# Undergraduate Student: Arturo Burgos
# Professor: João Rodrigo Andrade
# Federal University of Uberlândia - UFU, Fluid Mechanics Laboratory - MFLab, Block 5P, Uberlândia, MG, Brazil

using Printf # Just the same as import a module
using Statistics
using CPUTime

println("Choose a N value: ")
N = chomp(readline()) # Here I get the char from the user
n = parse(Int32,N)  # Here I convert the char into an integer(32bits)

A = rand(1:200,n,n);
B = rand(1:200,n,n);

D = A*B; # This is the vectorized way for this problem, like MATLAB
C = A.*B; # This multiplies point by point

R = zeros(n,n) # Here I create a zero matrix just to get the multiply of A and B

# println("\n\nThe D Matrix is: \n\n",D) # Compare with low n values 


function multivec(mtx1,mtx2,mtx3)

    mtx3 = mtx1*mtx2

    return mtx3 # Returns a value to the variable
end


R = multivec(A,B,R);
# println("\n\nThe R Matrix is: \n\n",R) # Compare with low n values

println("\n")

@time @CPUtime multivec(A,B,R)
# Test done!